### PR TITLE
[5.7] cherry-pick table filter

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Cherry pick [table filter](https://github.com/facebook/rocksdb/pull/2265).

### CODE REVIEW

**Comment by huachaohuang** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
@UncP PTAL

File: include/rocksdb/options.h [hunk 1]
```diff
@@ -1063,6 +1063,14 @@ struct ReadOptions {
   // Default: false
   bool ignore_range_deletions;
 
+  // A callback to determine whether relevant keys for this scan exist in a
+  // given table based on the table's properties. The callback is passed the
```
Comment (UncP) [inline include/rocksdb/options.h]: `is passed the` seems to be a grammar error.

Comment (zhangjinpeng1987) [reply include/rocksdb/options.h]: That does't make sense, we just make it consistent with upstream change.

**Comment by UncP** [comment]
LGTM

**Comment by siddontang** [comment]
please fix the conflict

**Pull request merged**